src/modules/submodules/consensus.py

```diff
@@ -0,0 +1,362 @@
+                if self.is_main_data_submitted(latest_blockstamp):
+                    logger.info({'msg': 'Main data was submitted.'})
+                    return
+
+        if self.is_main_data_submitted(latest_blockstamp):
+            logger.info({'msg': 'Main data was submitted.'})
+            return
+
+        logger.info({'msg': f'Send report data. Contract version: [{self.CONTRACT_VERSION}]'})
+        # If data already submitted transaction will be locally reverted, no need to check status manually
+        self._submit_report(report_data, self.CONTRACT_VERSION)
+
+    def _get_latest_data(self) -> tuple[BlockStamp, MemberInfo]:
+        latest_blockstamp = self._get_latest_blockstamp()
+        logger.debug({'msg': 'Get latest blockstamp.', 'value': latest_blockstamp})
+
+        member_info = self.get_member_info(latest_blockstamp)
+        logger.debug({'msg': 'Get current member info.', 'value': member_info})
+        return latest_blockstamp, member_info
+
+    def _get_report_hash(self, report_data: tuple):
+        # The Accounting Oracle and Ejector Bus has same named method to report data
+        report_function_name = 'submitReportData'
+
+        report_function_abi = next(filter(lambda x: 'name' in x and x['name'] == report_function_name, self.report_contract.abi))
+
+        # First input is ReportData structure
+        report_data_abi = report_function_abi['inputs'][0]['components']
+
+        # Transform abi to string
+        report_str_abi = ','.join(map(lambda x: x['type'], report_data_abi))
+
+        # Transform str abi to tuple, because ReportData is struct
+        encoded = encode([f'({report_str_abi})'], [report_data])
+
+        report_hash = self.w3.keccak(encoded)
+        logger.info({'msg': 'Calculate report hash.', 'value': report_hash})
+        return report_hash
+
+    def _send_report_hash(self, blockstamp: ReferenceBlockStamp, report_hash: bytes, consensus_version: int):
+        consensus_contract = self._get_consensus_contract(blockstamp)
+
+        tx = consensus_contract.functions.submitReport(blockstamp.ref_slot, report_hash, consensus_version)
+
+        self.w3.transaction.check_and_send_transaction(tx, variables.ACCOUNT)
+
+    def _submit_report(self, report: tuple, contract_version: int):
+        tx = self.report_contract.functions.submitReportData(report, contract_version)
+
+        self.w3.transaction.check_and_send_transaction(tx, variables.ACCOUNT)
+
+    def _get_latest_blockstamp(self) -> BlockStamp:
+        root = self.w3.cc.get_block_root('head').root
+        bs = build_blockstamp(self.w3.cc, root)
+        logger.debug({'msg': 'Fetch latest blockstamp.', 'value': bs})
+        return bs
+
+    @lru_cache(maxsize=1)
+    def _get_slot_delay_before_data_submit(self, blockstamp: BlockStamp) -> int:
+        """Returns in slots time to sleep before data report."""
+        member = self.get_member_info(blockstamp)
+        if member.is_submit_member:
```

Comment:
`member.is_submit_member` → `member.is_submitter` maybe?

we treat `member` as a committee participant